packages/teleport-plugin-stencil-base-component/src/index.ts

```diff
@@ -0,0 +1,97 @@
+import { createClassDeclaration, createComponentDecorator } from './utils'
+import createJSXSyntax from '@teleporthq/teleport-shared/dist/cjs/node-handlers/node-to-jsx'
+
+import { ComponentPluginFactory, ComponentPlugin } from '@teleporthq/teleport-types'
+
+import {
+  DEFAULT_COMPONENT_CHUNK_NAME,
+  DEFAULT_IMPORT_CHUNK_NAME,
+  DEFAULT_COMPONENT_DECORATOR_CHUNK_NAME,
+  STENCIL_CORE_DEPENDENCY,
+} from './constants'
+
+interface StencilPluginConfig {
+  componentChunkName: string
+  componentDecoratorChunkName: string
+  importChunkName: string
+}
+
+export const createPlugin: ComponentPluginFactory<StencilPluginConfig> = (config) => {
+  const {
+    componentChunkName = DEFAULT_COMPONENT_CHUNK_NAME,
+    componentDecoratorChunkName = DEFAULT_COMPONENT_DECORATOR_CHUNK_NAME,
+    importChunkName = DEFAULT_IMPORT_CHUNK_NAME,
+  } = config || {}
+
+  const stencilComponentPlugin: ComponentPlugin = async (structure) => {
+    const { uidl, dependencies } = structure
+    const { stateDefinitions = {}, propDefinitions = {} } = uidl
+
+    dependencies.Component = STENCIL_CORE_DEPENDENCY
+    dependencies.h = STENCIL_CORE_DEPENDENCY
+
+    if (Object.keys(propDefinitions).length > 0) {
+      dependencies.Prop = STENCIL_CORE_DEPENDENCY
+    }
+
+    if (Object.keys(stateDefinitions).length > 0) {
+      dependencies.State = STENCIL_CORE_DEPENDENCY
+    }
+
+    // We will keep a flat mapping object from each component identifier (from the UIDL) to its correspoding JSX AST Tag
+    // This will help us inject style or classes at a later stage in the pipeline, upon traversing the UIDL
+    // The structure will be populated as the AST is being created
+    const nodesLookup = {}
+    const jsxParams = {
+      propDefinitions,
+      stateDefinitions,
+      nodesLookup,
+      dependencies,
+    }
+
+    const jsxOptions = {
+      dynamicReferencePrefixMap: {
+        prop: 'this',
+        state: 'this',
+        local: '',
+      },
+      useHooks: false,
+    }
+
+    const jsxTagStructure = createJSXSyntax(uidl.node, jsxParams, jsxOptions)
+    const exportAST = createClassDeclaration(
+      uidl.name,
+      propDefinitions,
+      stateDefinitions,
+      jsxTagStructure
+    )
+
+    const decoratorAST = createComponentDecorator(uidl.name)
+
+    structure.chunks.push({
+      type: 'js',
+      name: componentDecoratorChunkName,
```

Comment:
we're currently matching js/html/string as types of chunks in the builder with generators (babel/generator, hast-to-html, etc.). we might use different types at some point, it might be getting confusing now, but should stay `js`

Reply:
Yeah, I got that. But currently, the output file is being generated as **js**, we can't have typings in a **js** file right!

Reply:
this is not the file extension, it's just the type of chunk which is handled by a specific generator (ast to string or hast to string). stencil file should be generated in tsx, but we'll check that when we do a project generator